### DISCUSSION TITLE
machine.reset() does nothing?

### DISCUSSION THREAD

**Discussion by gitcnd** [micropython/micropython#12856] [Hardware & Peripherals]
My web REPL simply disconnects when I machine.reset() (using latest firmware on ESP-WROOM32)

How does the kernel handle a reset?  Does the ESP32 hardware itself even have any feature which actually can reset itself ?

**Comment by jimmo** [comment]
This is to be expected. `machine.reset()` triggers a hard reset, it's roughly equivalent to asserting the RST pin on the microcontroller. All memory is cleared and the wifi is reset.

**Comment by gitcnd** [comment]
The [board](https://github.com/NoosaHydro/2.4inch_ESP32-2432S024/tree/main) does *not* reset, and neither do any of the peripherals (text stays on the LCD, RGB led keeps its color, ...).  I'm using https://micropython.org/resources/firmware/ESP32_GENERIC-OTA-20231005-v1.21.0.bin

_*How*_ does the kernel handle a reset? Does the ESP32 hardware itself even _have_ any feature which actually can reset itself ?

**Reply by jimmo** [reply]
I'm not really sure what you mean...

If by "kernel" you mean MicroPython (and to some extent the ESP IDF), when you call machine.reset() it's literally like someone pressed the reset button. All on-chip peripherals are reset, and the cpu just boots as if it was powered on. Anything else on the board is largely oblivious to this (until they are sent commands from the cpu).

It's kind of like, if you reset your desktop PC, your printer doesn't reset. In the case of an embedded system, the display and the RGB led are all stateful, they will just keep showing whatever in in their internal buffer.

If you want to clear the display/led on reset, then your program needs to do that at startup.

**Comment by robert-hh** [comment]
The devices I have here restart if I call machine.reset(). The code for machine.reset() calls esp_restart(). If you connect to the board by USB using a simple terminal program, and call:

```
import machine
machine.reset()
```

then you should set the usual boot sequence. Any attached peripheral is not affected by this as long as power stays active, unless reset by your code.

**Comment by shariltumin** [comment]
You can test what reset was done on the board recently:

```py
>>> import machine
>>> machine.reset_cause()
2
```

The numbers returned by ```machine.reset_cause()``` are defined as follows:

```py
>>> machine.PWRON_RESET
1
>>> machine.HARD_RESET
2
>>> machine.WDT_RESET
3
>>> machine.DEEPSLEEP_RESET
4
>>> machine.SOFT_RESET
5
>>> CTRL+D
MPY: soft reboot
MicroPython v1.22.0-preview.63.g2fda94c28.dirty on 2023-11-02; ESP32 KAKI5 with ESP32
>>> import machine
>>> machine.reset_cause()
5
```
